src/transport/UnauthenticatedSessionTable.h

```diff
@@ -0,0 +1,204 @@
+    {
+        UnauthenticatedSession * result = nullptr;
+        mEntries.ForEachActiveObject([&](UnauthenticatedSession * entry) {
+            if (MatchPeerAddress(entry->GetPeerAddress(), address))
+            {
+                result = entry;
+                return false;
+            }
+            return true;
+        });
+        return result;
+    }
+
+    /**
+     * Get a peer given the peer id. If the peer doesn't exist in the cache, allocate a new entry for it.
+     *
+     * @return the peer found or allocated, nullptr if not found and allocate failed.
+     */
+    CHECK_RETURN_VALUE
+    UnauthenticatedSession * FindOrAllocateEntry(const PeerAddress & address)
+    {
+        UnauthenticatedSession * result = FindEntry(address);
+        if (result != nullptr)
+            return result;
+
+        CHIP_ERROR err = AllocEntry(address, result);
+        if (err == CHIP_NO_ERROR)
+        {
+            return result;
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+
+    /// Mark a session as active
+    void MarkSessionActive(UnauthenticatedSession & entry) { entry.SetLastActivityTimeMs(mTimeSource.GetCurrentMonotonicTimeMs()); }
+
+    /// Allows access to the underlying time source used for keeping track of connection active time
+    Time::TimeSource<kTimeSource> & GetTimeSource() { return mTimeSource; }
+
+private:
+    UnauthenticatedSession * FindLeastRecentUsedEntry()
+    {
+        UnauthenticatedSession * result = nullptr;
+        uint64_t oldestTimeMs           = std::numeric_limits<uint64_t>::max();
+
+        mEntries.ForEachActiveObject([&](UnauthenticatedSession * entry) {
+            if (entry->GetReferenceCount() == 0 && entry->GetLastActivityTimeMs() < oldestTimeMs)
+            {
+                result       = entry;
+                oldestTimeMs = entry->GetLastActivityTimeMs();
+            }
+            return true;
+        });
+
+        return result;
+    }
+
+    static bool MatchPeerAddress(const PeerAddress & a1, const PeerAddress & a2)
+    {
+        if (a1.GetTransportType() != a2.GetTransportType())
+            return false;
+
+        switch (a1.GetTransportType())
+        {
+        case Transport::Type::kUndefined:
+            return false;
+        case Transport::Type::kUdp:
+        case Transport::Type::kTcp:
+            return a1.GetIPAddress() == a2.GetIPAddress() && a1.GetPort() == a2.GetPort() &&
+                (a1.GetInterface() == INET_NULL_INTERFACEID || a2.GetInterface() == INET_NULL_INTERFACEID ||
+                 a1.GetInterface() == a2.GetInterface());
+        case Transport::Type::kBle:
+            // TODO: complete BLE address comparation
+            return true;
+        }
+
+        return false;
+    }
+
+    static constexpr uint64_t kMinimalActivityTimeMs = 30000;
+    Time::TimeSource<Time::Source::kSystem> mTimeSource;
+    BitMapObjectPool<UnauthenticatedSession, kMaxConnectionCount> mEntries;
```

Comment:
For more capable nodes, it would be so much better not to to be forced to determine `CHIP_CONFIG_UNAUTHENTICATED_CONNECTION_POOL_SIZE` at compile time.  Think of a larger controller that is booting up and wants to reestablish CASE sessions with all of the nodes on a fabric.  This places a hard constraint on the number of CASE sessions we can established at a time.

Add to this the fact CASE establishment is time consuming for constrained peers and you have a major scaling problem.  At 5 seconds per peer and with `CHIP_CONFIG_UNAUTHENTICATED_CONNECTION_POOL_SIZE` of 4, it would take > 10 minutes to establish 500 connections.

How do we get to a place where the UnauthenticatedSession pool can be arbitrarily large on nodes where this makes sense?

Reply:
@yufengwangca is working on the pool implementation to support dynamic allocation, on non-memory-intense platforms. We will migrate to the new implementation when it is done.

Reply:
Thanks!